ocrd/ocrd/mets_server.py

```diff
@@ -0,0 +1,253 @@
+            loctype (string): ``@LOCTYPE`` of this ``mets:file``
+            local_filename (): ignored
+            url (string): ``@xlink:href`` of this ``mets:file``
+            ID (string): ``@ID`` of this ``mets:file``
+        """
+        self.ID = ID
+        self.mimetype = mimetype
+        self.url = url
+        self.loctype = loctype
+        self.pageId = pageId
+        self.fileGrp = fileGrp
+
+class ClientSideOcrdMets():
+    """
+    Replacement for :py:class:`ocrd_models.ocrd_mets.OcrdMets` with overrides for
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.find_files`,
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.find_all_files`, and
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.add_file` to query via HTTP a
+    :py:class:`ocrd.mets_server.OcrdMetsServer`.
+    """
+
+    def __init__(self, host, port, socket):
+        self.log = getLogger('ocrd.mets_client.%s' % ('uds' if socket else 'tcp'))
+        if socket:
+            self.url = f'http+unix://{socket.replace("/", "%2F")}'
+            self.session = requests_unixsocket_session()
+        else:
+            self.url = f'http://{host}:{port}'
+            self.session = requests_session()
+
+    def find_files(self, **kwargs):
+        r = self.session.request('GET', self.url, params={**kwargs})
+        for f in r.json()['files']:
+            yield ClientSideOcrdFile(None, ID=f['file_id'], pageId=f['page_id'], fileGrp=f['file_grp'], url=f['url'], mimetype=f['mimetype'])
+
+    def find_all_files(self, *args, **kwargs):
+        return list(self.find_files(*args, **kwargs))
+
+    def add_agent(self, *args, **kwargs):
+        return self.session.request('POST', f'{self.url}/agent', data=OcrdAgentModel.create(**kwargs))
+
+    @property
+    def file_groups(self):
+        return self.session.request('GET', f'{self.url}/file_groups').json()['file_groups']
+
+    @deprecated_alias(pageId="page_id")
+    @deprecated_alias(ID="file_id")
+    def add_file(self, file_grp, content=None, file_id=None, url=None, mimetype=None, page_id=None, **kwargs):
+        r = self.session.request(
+            'POST',
+            self.url,
+            data=OcrdFileModel.create(
+                file_id=file_id,
+                file_grp=file_grp,
+                page_id=page_id,
+                mimetype=mimetype,
+                url=url).dict(),
+            files={'data': content}
+        )
+
+    def save(self):
+        self.session.request('PUT', self.url)
+
+
+#
+# Server
+#
+
+class OcrdMetsServer():
+
+    def __init__(self, workspace, host, port, socket):
+        self.workspace = workspace
+        self.host = host
+        self.port = port
+        self.socket = socket
+        self.log = getLogger('ocrd.workspace_client')
+
+
+    def startup(self):
+
+        # XXX HACK 
+        # circumventing dependency injection like this is bad and
+        # needs to be refactored once it's all runnign
+        workspace = self.workspace
```

Comment:
I don't understand this hack, why is it neccesarry. I removed it and used `self.workspace` instead previously `workspace`. After that I didn't noticed a difference (but did just briefly tested adding a file)

Reply:
The `workspace = self.workspace` is not the hack, that is just for convenience, it makes no difference.

The hack IMHO is that all the defined endpoints refer to a "global" variable `workspace` that is initiated before the routes are even defined. From what I understand about fastapi, the right way is to use dependency injection and module level functions. But your approach with subclassing FastAPI might be a good solution.

Reply:
ok. I think/thought "global" variables is how this kind of things are done in fastapi. So I would assume if not subclassing this would be ok or maybe do `app.workspace = self.workspace`. But anyway my question is resolved